File: Chapter 2/EpsilonGreedyBandit.m
classdef Bandit
    properties
        k
        epsilon
        q_true
        q_estimated
        action_count
    end
    
    methods
        function obj = Bandit(k_arms, epsilon)
            obj.k = k_arms;
            obj.epsilon = epsilon;
            obj.q_true = randn(1, k_arms);
            obj.q_estimated = zeros(1, k_arms);
            obj.action_count = zeros(1, k_arms);
        end
        
        function action = act(obj)
            if rand() < obj.epsilon
                action = randi(obj.k);
            else
                [~, action] = max(obj.q_estimated);
            end
        end
        
        function reward = step(obj, action)
            reward = randn() + obj.q_true(action);
            obj.action_count(action) = obj.action_count(action) + 1;
            obj.q_estimated(action) = obj.q_estimated(action) + ...
                (reward - obj.q_estimated(action)) / obj.action_count(action);
        end
    end
end

% Example usage
bandit = Bandit(10, 0.1);
for i = 1:1000
    action = bandit.act();
    reward = bandit.step(action);
end

File: Chapter 2/EpsilonGreedyBandit.py
import numpy as np

class Bandit:
    def __init__(self, k_arms, epsilon):
        self.k = k_arms
        self.epsilon = epsilon
        self.q_true = np.random.randn(k_arms)  # True values
        self.q_estimated = np.zeros(k_arms)    # Estimated values
        self.action_count = np.zeros(k_arms)   # Count of actions taken

    def act(self):
        if np.random.rand() < self.epsilon:
            return np.random.choice(self.k)  # Explore
        else:
            return np.argmax(self.q_estimated)  # Exploit

    def step(self, action):
        reward = np.random.randn() + self.q_true[action]
        self.action_count[action] += 1
        self.q_estimated[action] += (reward - self.q_estimated[action]) / self.action_count[action]
        return reward

# Example usage
bandit = Bandit(10, 0.1)
for _ in range(1000):
    action = bandit.act()
    reward = bandit.step(action)
